Add soc em32f967

ElanJohnnyChuang opened the pull request:
soc: elan: add em32f967 SoC and 32f967_dv board support

This pull request adds initial support for the Elan em32f967 SoC and the
32f967_dv development board, based on ARM Cortex-M4.

The current implementation provides:
- Basic SoC structure under `soc/elan/em32f967`
- Device tree and Kconfig setup for the SoC and board
- Devicetree bindings for clock, GPIO, and pinctrl
- Remove `hal_elan` module and keep all code in-tree

At this stage, only the `samples/hello_world` application builds successfully.
Peripheral drivers (UART, GPIO, etc.) are not yet included, so runtime output
is not expected.

#### Build instructions
```bash
west build -p always -b 32f967_dv samples/hello_world
```

ElanJohnnyChuang committed:
soc: add Elan em32f967 minimal SoC support

Add initial support for the Elan em32f967 SoC based on ARM Cortex-M4.
Minimal SoC definition for upstream:
- soc.h, Kconfig, and CMakeLists.txt provided for completeness
- This PR does not implement drivers or UART functionality

Signed-off-by: Johnny Chuang <johnny.chuang@emc.com.tw>

ElanJohnnyChuang committed:
board: add '32f967_dv' development board (minimal build support)

Minimal board definition for the Elan em32f967 SOC:
- Includes DTS, YAML, and Kconfig files
- Provides board.cmake with misc-flasher for CI/build
- Actual flashing is not included in this PR

Signed-off-by: Johnny Chuang <johnny.chuang@emc.com.tw>

nordicjm commented on soc/elan/em32f967/elan_em32.h:
```diff
@@ -0,0 +1,228 @@
+	HCLKF_GHM_M2D = 0x2b,
+	PCLKG_SSP1 = 0x30,
+	PCLKG_ALL = 0xffff
+} CLKGatingSwitch;
```
zephyr is snake case not camel case

ElanJohnnyChuang commented on dts/arm/elan/em32fxxx.dtsi:
```diff
@@ -0,0 +1,184 @@
+		reg = <0x20000000 DT_SIZE_K(448)>;
+	};
+
+    clocks {
```
This draft PR update is mainly for me to get familiar with the workflow. I’m aware that the current code still has many aspects that don’t meet the requirements. I will address those based on your suggestions and the issues reported by the CI pipeline. After the necessary fixes and optimizations, I’ll force‑push the changes to this PR and then convert the draft PR into a regular PR for your review. Thank you.

ElanJohnnyChuang committed:
dt-bindings: elan em32f967 – clock, gpio, pinctrl, pwm

Add minimal DT binding headers and matching YAML files for the
Elan em32f967 SoC. The new definitions are required to build the
sample board and follow the upstream style.

Signed-off-by: Johnny Chuang <johnny.chuang@emc.com.tw>

ElanJohnnyChuang commented on soc/elan/em32f967/Kconfig.defconfig:
```diff
@@ -0,0 +1,15 @@
+	default 64
+
+config SYS_CLOCK_HW_CYCLES_PER_SEC
+	default 60000000
```
Updated the config as suggested.

ElanJohnnyChuang commented on soc/elan/em32f967/soc.h:
```diff
@@ -0,0 +1,19 @@
+#include "em32f967.h"
+#include "soc_967.h"
+
+extern const struct device __device_dts_ord_DT_CHOSEN_zephyr_cortex_m_idle_timer_ORD;
```
Thanks for pointing it out, I forgot to remove this.

nordicjm commented on boards/elan/32f967_dv/32f967_dv.dts:
```diff
@@ -0,0 +1,157 @@
+		ec_rw: partition@10024000 {
+			label = "ec-rw";
+			reg = <0x10024000 DT_SIZE_K(392)>;
+		};
```
how does this work, I assume the zephyr application should be build for one of these and the rest ignored? If so, the chosen node for the partition should be set then `CONFIG_USE_DT_CODE_PARTITION` should be selected in the board `Kconfig` file

ElanJohnnyChuang commented on soc/elan/em32f967/soc.c:
```diff
@@ -0,0 +1,12 @@
+#include <zephyr/init.h>
+
+/* Minimal SoC init for reviewer and build */
+static int soc_elan_em32f967_init(void)
```
Currently a minimal init for build verification.
We may extend or relocate it when integrating clock/pinctrl initialization.

ElanJohnnyChuang commented:
Hi Reviewers, 
I've updated the PR and included changes to the HAL. Could you please approve the CI tests when you have a chance? Thanks!

nordicjm commented on west.yml:
```diff
@@ -168,6 +168,10 @@ manifest:
         - hal
+    - name: hal_elan
+      path: modules/hal/elan
+      revision: 50e4bb7889193687fc63b5cbcb11aa383ac3aa92
```
manifest update must come before things that need it

nordicjm commented on boards/elan/32f967_dv/Kconfig.defconfig:
```diff
@@ -0,0 +1,21 @@
+	default 4096
+
+config IDLE_STACK_SIZE
+	default 1024
```
why are these needed

ElanJohnnyChuang commented on soc/elan/em32f967/soc.c:
```diff
@@ -0,0 +1,17 @@
+#include <zephyr/init.h>
+
+/* Minimal SoC init for reviewer and build */
+static int soc_elan_em32f967_init(void)
```
Thanks for the suggestion. I’ve removed soc/elan/em32f967/soc.c since it only provided a no‑op SYS_INIT and added no functional initialization.
If/when we need early SoC setup (clocks/PLL, VTOR, errata, MPU/Cache), I’ll add a proper SYS_INIT back with the right phase/priority.

ElanJohnnyChuang commented on dts/bindings/clock/elan,em32-ahb.yaml:
```diff
@@ -0,0 +1,27 @@
+# Copyright (c) 2025 Elan Microelectronics Corp..
+# SPDX-License-Identifier: Apache-2.0
+
```
Thanks for pointing this out. I’ve reordered the commits so that the DT bindings (elan,em32-ahb.yaml) land before the board addition, since the board depends on them. The new order is:

- west manifest update
- DT bindings (clock: elan,em32-ahb.yaml)
- SoC / DTS updates
- Board (32f967_dv)

I verified a pristine build after the rebase.

nordicjm commented on MAINTAINERS.yml:
```diff
@@ -5487,6 +5487,15 @@ West:
+"West project: hal_elan":
+  status: maintained
+  maintainers:
+    - galak
```
I highly doubt galak will be maintaining this...

ElanJohnnyChuang replied:
Thanks for pointing that out — you’re absolutely correct that galak is not the actual maintainer for this HAL.
I initially added galak based on CI requirements and previous guidance, but I understand that maintainers: should reflect the real code owners recognized by the project.

I am submitting this HAL on behalf of Elan, under my personal GitHub account (@ElanJohnnyChuang). I can take over maintenance. Is it acceptable to list me in maintainers:? If Zephyr requires TSC nomination or approval for vendor maintainers, I’m happy to follow the process.

Proposed update:
```
"West project: hal_elan":
  status: maintained
  maintainers:
    - ElanJohnnyChuang
```

Please let me know if I should instead use the org account (@elantw‑zephyr) or if further steps are needed.

nordicjm replied:
You can maintain it, you need to go to github issues and create a new one of type external component integration and fill the details in with a link to the repo and if approved, a repo will be created into of zephyr which you can use

fabiobaltieri reviewed (commented):
Would be nice to have a board picture as well, just for reference

ElanJohnnyChuang commented on boards/elan/32f967_dv/doc/index.rst:
```diff
@@ -0,0 +1,42 @@
+.. _elan_32f967_dv:
```
@kartben Thanks!

I’ve updated `boards/elan/32f967_dv/doc/index.rst` to follow the board template:
If you’d like any other adjustments to match the template more strictly, I’ll update right away.

ElanJohnnyChuang commented:
Hi @kartben,
Hope you had a great holiday!
I’ve updated the PR to follow the board template as suggested. Two other reviewers have approved the changes.
When you get a chance, could you please re-review? Let me know if anything else needs tweaking.
Thanks!
